### PR TITLE
Messaging System

### PR DESCRIPTION
**Pull request by njgrisafi**
Adds the private messaging system to the `wtf.core` and `wtf.api`

This is the basic implementation to do the following:
- Send a message
- Reply to a message
- Retrieve a message
- Retrieve a message replies
- Retrieve messages for an recipient

Part of #55

### CODE REVIEW

File: wtf/api/routes.py [hunk 1]
```diff
@@ -324,3 +329,82 @@ def get_armor_by_id(armor_id):
...
     return jsonify({'armor': armor.transform(existing_armor)}), 200
+
+
+@BLUEPRINT.route('/messages', methods=['GET'])
+def get_messages_query_request():
+    '''Handle message retrieval by recipient id.
+
+    $ curl \
+        --request GET \
+        --url http://localhost:5000/api/messages?recipient=<recipient_id>?status=<status> \
```
Comment (mrpudn) [inline wtf/api/routes.py]: I think this should be (`&`):
```
...?recipient=<recipient_id>&status=<status>
```

File: wtf/api/routes.py [hunk 2]
```diff
@@ -324,3 +329,82 @@ def get_armor_by_id(armor_id):
...
+            "recipients": ["..."],
+        }'
+    '''
+    body = get_json_body()
+    message = messages.save(messages.create(
+        subject=body.get('subject'),
+        body=body.get('body'),
+        recipients=body.get('recipients')
+    ))
+    return jsonify({'message': message}), 200
```
Comment (mrpudn) [inline wtf/api/routes.py]: I think this should be a `201 CREATED`

File: wtf/testing.py [hunk 3]
```diff
@@ -48,10 +48,12 @@ def post(self, path='', **kwargs):
     def get(self, path='', **kwargs):
         '''Send a GET request'''
         path = '%s%s' % (self.root_path, path)
+        query_string = kwargs.get('query_string', {})
```
Comment (mrpudn) [inline wtf/testing.py]: Mind if we shorten this to like `query`? What do you think?

File: wtf/core/messages_test.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    expected = 'foobar'
+    messages.REPO['by_recipient'][TEST_DATA['id']] = expected
+    actual = messages.find_by_recipient(TEST_DATA['id'])
+    assert expected == actual
+
+
+def test_find_message_by_recipient_not_found():
+    expected = 'Recipient not found'
+    with pytest.raises(NotFoundError) as e:
+        messages.find_by_recipient(TEST_DATA['id'])
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: not sure if I'm reading this correctly: `TEST_DATA['id']` should be the message ID, right?

File: wtf/core/messages_test.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    assert set(expected).issubset(actual)
+
+@patch('wtf.core.messages.find_by_recipient')
+def test_get_recipient_messages(mock_find_by_recipient):
+    expected = ['foobar']
+    messages.REPO['by_id'][TEST_DATA['id']] = expected[0]
+    mock_find_by_recipient.return_value = [
+        {'message': TEST_DATA['id'], 'status': 'unread'}
+    ]
+    actual = messages.get_recipient_messages()
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: How does this know which recipient to fetch methods for? Or is this just all messages?

File: wtf/core/messages_test.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+def test_validate_message_missing_fields():
+    expected = [
+        'Missing required field: subject',
+        'Missing required field: body',
+        'Missing required field: copies'
+    ]
+    with pytest.raises(ValidationError) as e:
+        messages.validate({})
+    actual = e.value.errors
+    assert set(expected).issubset(actual)
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: Hey, regarding the validation stuff we talked about in gitter: don't worry about it for this PR. I'll probably come back and fix them all in a single PR later. :+1:

File: wtf/core/messages_test.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        messages.save({})
+    assert not messages.REPO['by_id'].values()
+    assert not messages.REPO['by_recipient'].values()
+
+
+def test_validate_message():
+    messages.validate({
+        'id': TEST_DATA['id'],
+        'parent': None,
+        'sender': None,
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: Sender should always have a value, right?

File: wtf/core/messages_test.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        'subject': TEST_DATA['subject'],
+        'body': TEST_DATA['body'],
+        'copies': []
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: Copies shouldn't be empty, right?

File: wtf/core/messages_test.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                'status': 'unread'
+            },
+            {
+                'deleted_at': None,
+                'read_at': None,
+                'message': TEST_DATA['id'],
+                'recipient': 'bar',
+                'status': 'unread'
+            }
+        ]
```
Comment (mrpudn) [inline wtf/core/messages_test.py]: I think my explanation of message copies wasn't very clear. I meant for them to be like whole separate entities, not embedded objects.

Imagine how you'd represent them in a relational database fashion. `messages` table, related to `message_copies` table by `messages.id` equal to `message_copies.message`.

Comment (mrpudn) [reply wtf/core/messages_test.py]: well actually, they do **become** embedded objects, but only after the `transform()` part.. actually still not totally clear on that myself

File: wtf/core/messages_test.py (hunk 7 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: For now we don't associate anything with sender so that's why I left it as None. I can add in the test case just so we have it their for the future

File: wtf/testing.py (hunk 3 above)
Comment (njgrisafi) [reply wtf/testing.py]: Yup I like that! I'll make the change

File: wtf/core/messages_test.py (hunk 4 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: yup that's suppose to come from messages test data. Good catch!

File: wtf/core/messages_test.py (hunk 8 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: I guess that makes sense. You can't send a message without a recipient. I'll add that check in the validation

File: wtf/core/messages_test.py (hunk 9 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: Ahh okay yeah I didn't think embedding the same data twice made much sense.

Do you want me to make a new file `core/message_copies.py` or keep message copies in the `core/messages.py`?

I think you're saying you just want me to add another repo in `core/messages.py` for `message_copies_by_message_id` . Then when a user requests `/message/<message_id>`, we fetch the message copies, then transform the data to include the copies and return the transformed data to the user.

Or do we only want to return the array of recipient addresses in `message.copies`?

Sorry if this is confusing, I'm bit lost right now haha

File: wtf/core/messages_test.py (hunk 4 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: Yup just made the fix to use recipients

File: wtf/core/messages_test.py (hunk 7 above)
Comment (mrpudn) [reply wtf/core/messages_test.py]: Oh, I thought it would have the character ID of the character sending the message?

File: wtf/core/messages_test.py (hunk 9 above)
Comment (mrpudn) [reply wtf/core/messages_test.py]: haha no worries - we can probably remove the message copies stuff and tackle that in a second PR if that's not too difficult. I'll put some thoughts into the issue for this.

Comment (njgrisafi) [reply wtf/core/messages_test.py]: I think we can get it done in this PR. Your last update on the issue help clear things up

File: wtf/core/messages_test.py (hunk 7 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: @mrpudn ahh I thought that it was eventually going to be the account ID. Once we have the `auth` in place then we should be able to populate the `sender` value right?

File: wtf/core/messages_test.py (hunk 5 above)
Comment (njgrisafi) [reply wtf/core/messages_test.py]: I should've used the recipient ID. I updated this with my last commits

**Comment by njgrisafi** [comment]
@mrpudn Updated this PR with the following:

- Refactored message replies into separate repo
- Added transform method for message and message_copies
- Added retrieving replies for a message

I'm thinking we should create create a `save_copies` method because the message `save` method is getting large and maybe out of it's scope.

Let me know if you catch anything else I should change. I think some of the naming and error messages should be refactored at some point too.

**Comment by mrpudn** [comment]
Yup, go ahead with `save_copy()` or `save_copies()` (or both, even) - there's a `save_recipe()` and a `save()` method in `weapons` and `armor`, and I'm about to add a `save_token()` to the `accounts` module, so totally fits the pattern.

File: wtf/api/routes.py [hunk 10]
```diff
@@ -25,6 +25,11 @@ def get_json_body():
     return body
 
 
+def get_query_args():
+    '''Get the request query args.'''
+    return request.args
```
Comment (mrpudn) [inline wtf/api/routes.py]: Any reason there's a method for this? :P

Seem like we could just use `request.args` on its own - were you thinking we might add something to this in the future?

File: wtf/api/routes.py [hunk 11]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
     return jsonify({'armor': armor.transform(existing_armor)}), 200
+
+
+@BLUEPRINT.route('/messages', methods=['GET'])
+def get_messages():
+    '''Handle message retrieval by recipient id.
+
+    $ curl \
+        --request GET \
+        --url http://localhost:5000/api/messages?recipient=<recipient_id>&status=<status> \
```
Comment (mrpudn) [inline wtf/api/routes.py]: We'll have to figure out a better way to represent this in the future (probably via `--data-urlencode`), since there could be a lot of potential query args one could use. (No need to tackle that in this PR, though)

File: wtf/api/routes.py [hunk 12]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+        --write-out "\n"
+    '''
+    args = get_query_args()
+    status = args.get('status')
+    recipient = args.get('recipient')
+    message_copies = messages.get_recipient_message_copies(recipient=recipient, status=status)
```
Comment (mrpudn) [inline wtf/api/routes.py]: How about maybe `find_by_recipient_id()` to match the convention elsewhere (`accounts.find_by_email()`, `characters.find_by_id()`, etc.)?

File: wtf/api/routes.py [hunk 13]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+    return jsonify({'messages': messages.transform_copies(message_copies)}), 200
```
Comment (mrpudn) [inline wtf/api/routes.py]: I don't think we've dealt with multiple things getting transformed - how about `transform_all()` since this should transform multiple raw messages and add their message copies.

What gets passed in here should be an array of messages. The `transform_all()` will fetch the message copies for you.

File: wtf/api/routes.py [hunk 14]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+            "recipients": ["..."],
+        }'
+    '''
+    body = get_json_body()
+    message = messages.save(messages.create(
+        subject=body.get('subject'),
+        body=body.get('body'),
+        recipients=body.get('recipients')
+    ))
+    return jsonify({'message': message}), 201
```
Comment (mrpudn) [inline wtf/api/routes.py]: This should be a `transform()`-ed message

File: wtf/api/routes.py [hunk 15]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+
+
+@BLUEPRINT.route('/messages/<message_id>', methods=['GET'])
+def get_message_by_id(message_id):
```
Comment (mrpudn) [inline wtf/api/routes.py]: `find_by_id` (since the module is already called `messages`, no need to add it here)

File: wtf/api/routes.py [hunk 16]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+
+    $ curl \
+        --request POST \
+        --url http://localhost:5000/api/messages/<message_id>/replies \
+        --header "Content-Type: application/json" \
+        --write-out "\n" \
+        --data '{
+            "subject": "...",
+            "body": "...",
+            "recipients": ["..."],
```
Comment (mrpudn) [inline wtf/api/routes.py]: no comma here :)

File: wtf/api/routes.py [hunk 17]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+    '''
+    messages.find_by_id(message_id)
+    body = get_json_body()
+    message = messages.save(messages.create(
+        subject=body.get('subject'),
+        body=body.get('body'),
+        recipients=body.get('recipients'),
+        parent=message_id
+    ))
+    return jsonify({'message': messages.transform(message)}), 201
```
Comment (mrpudn) [inline wtf/api/routes.py]: You know, for this route, I think we can reuse the `create_message()` handler and make `message_id` optional (i.e. `create_message(message_id=None)`.

See: https://stackoverflow.com/questions/14023664/flask-url-route-route-several-urls-to-the-same-function

Pretty cool, right? :)

File: wtf/api/routes.py [hunk 18]
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
...
+
+
+@BLUEPRINT.route('/messages/<message_id>/replies', methods=['GET'])
+def get_message_replies(message_id):
```
Comment (mrpudn) [inline wtf/api/routes.py]: What's this for again? Wouldn't you just use `GET /api/messages`? I haven't thought through fetching an entire reply chain yet

File: wtf/core/messages.py [hunk 19]
```diff
@@ -0,0 +1,208 @@
+'''
+wtf.core.messages
+
+Messages have the following properties:
+  * id: the message UUID (Universally Unique Identifier)
+  * body: the body of the message
+  * sender: the account id that send the messages
```
Comment (mrpudn) [inline wtf/core/messages.py]: I think this will be the character ID, not the account ID

File: wtf/core/messages.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+  * parent: the previous message id (None if no previous message)
+  * subject: the subject of the message
+  * copies: the recipients copies of the message
```
Comment (mrpudn) [inline wtf/core/messages.py]: This should be broken out into another entity with its own properties

File: wtf/core/messages.py [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+def create(**kwargs):
+    '''Create a message.
+
+    Messages have the following properties:
+        id: a UUID for the message
+        sender: a UUID for the sender
+        parent: the UUID for the previous message
+        subject: the subject of the message
+        body: the body of the message
+        recipients: the list of UUID for the recipients
```
Comment (mrpudn) [inline wtf/core/messages.py]: You don't need this (line 25 to line 31) since it's already in the module doc comment. Duplication :)

File: wtf/core/messages.py [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+    '''
+    message_id = kwargs.get('id')
+    body = kwargs.get('body')
+    subject = kwargs.get('subject', '(No Subject)')
+    recipients = kwargs.get('recipients')
+    recipients = recipients if recipients is not None else []
```
Comment (mrpudn) [inline wtf/core/messages.py]: you can replace these two lines with: 
```py
recipients = kwargs.get('recipients', [])
```

File: wtf/core/messages.py [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+        'id': message_id,
+        'sender': sender,
+        'parent': parent,
+        'subject': subject,
+        'body': body,
+        'copies': [],
+        'created_at': datetime.utcnow().isoformat()
+    }
+    for recipient in recipients:
+        message['copies'].append(create_copy(message, recipient=recipient))
```
Comment (mrpudn) [inline wtf/core/messages.py]: I think message copies is out of scope for this function and should not be nested like this. Let's continue discussion on this one in the issue thread (#92).

File: wtf/core/messages.py [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+    return message
+
+
+def create_copy(message, **kwargs):
+    '''Create a message copy
+
+    MessageCopies have the following properties:
```
Comment (mrpudn) [inline wtf/core/messages.py]: Not a big deal, but I think we should go with "Message Copies" instead of like a Java class "MessageCopies" to describe this.

File: wtf/core/messages.py [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+        message: a UUID for the original message
+        recipient: the UUID of the recipient
+        status: the status of the current message: unread, read, deleted, saved
+        read_at: the time the recipient read the mesaage
+        deleted_at: the time the recipient deleted the message
```
Comment (mrpudn) [inline wtf/core/messages.py]: Also, let's put this stuff in the module comment instead of here :)

File: wtf/core/messages.py [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+        'status': status,
+        'read_at': None,
+        'deleted_at': None
+    }
+
+
+def save(message):
+    '''Persist a message.
+
+    Creates a message
```
Comment (mrpudn) [inline wtf/core/messages.py]: No need for this line, you can just do `'''Persist a message.'''`

File: wtf/core/messages.py (hunk 22 above)
Comment (njgrisafi) [reply wtf/core/messages.py]: @mrpudn this wasn't working because the key exists and the value is set to None. This is because the calling function sets the default parameter to None. Since the key exist it will return the value (which could be none).

I guess we can avoid this by not setting the parameter if it is None.

https://www.tutorialspoint.com/python/dictionary_get.htm

File: wtf/core/messages.py (hunk 23 above)
Comment (njgrisafi) [reply wtf/core/messages.py]: Yup that's what I mentioned above. We can definitely break out this and the save logic for the message copies.

File: wtf/core/messages.py (hunk 19 above)
Comment (njgrisafi) [reply wtf/core/messages.py]: Okay I refactor that to be character ID

File: wtf/api/routes.py (hunk 18 above)
Comment (njgrisafi) [reply wtf/api/routes.py]: I thought it would be useful to retrieve all replies for a given parent message but I guess you can traverse the messages endpoint using parent value.

I'll remove this function.

File: wtf/api/routes.py (hunk 14 above)
Comment (njgrisafi) [reply wtf/api/routes.py]: Yeah I wasn't sure what to put here. Since you created the message I thought just to send back the raw inserted object. But yeah I'll switch it back to return the transformed value.

File: wtf/api/routes.py (hunk 13 above)
Comment (njgrisafi) [reply wtf/api/routes.py]: @mrpudn I think that's a great idea! I'll add that method and simplify this logic

File: wtf/api/routes.py (hunk 10 above)
Comment (njgrisafi) [reply wtf/api/routes.py]: Yeah no good reason, I'll go ahead and remove this. I was thinking to keep it consistent since many other function may be accessing the same data and we were doing something similar for the body.

File: wtf/api/routes.py (hunk 12 above)
Comment (njgrisafi) [reply wtf/api/routes.py]: @mrpudn I already added a method for `find_copies_by_recipient_id()`

This method `get_recipient_message_copies()`  calls `find_copies_by_recipient_id()` and filters the copies by the status (if status is set).

Should I just add the filtering logic in the `find_copies_by_recipient_id()` method?

**Comment by njgrisafi** [comment]
@mrpudn updated this branch. Let me know if you catch anything else you want me to change.

File: wtf/core/messages.py (hunk 22 above)
Comment (njgrisafi) [reply wtf/core/messages.py]: We can do `recipients = kwargs.get('recipients', [])` when we process the request and then call this function.

Comment (mrpudn) [reply wtf/core/messages.py]: oh gotcha, then you should be able to do this:
```py
recipients = kwargs.get('recipients') or []
```
Short and sweet :)

**Comment by mrpudn** [comment]
Hey, I'm going to go ahead and merge this. I'll probably create a follow up PR with a few tweaks.
